Lesson17/main.js

```diff
@@ -0,0 +1,120 @@
+const slideList = document.getElementById("js-slid-list");
+const slideWrapper = document.getElementById("js-slide-wrapper");
+
+const createElementWithClassName = (type, className) => {
+  const element = document.createElement(type);
+  element.className = className;
+  return element;
+}
+
+const addLoading = (parent) => {
+  const loading = createElementWithClassName("img", "loading");
+  loading.id = "js-loading";
+  loading.src = "./loading-circle.gif";
+  parent.appendChild(loading);
+}
+
+const removeLoading = () => document.getElementById("js-loading").remove();
+
+const addErrorMessage = (error, parent) => {
+  const errorMessage = createElementWithClassName("p", "error-message");
+  errorMessage.textContent = error;
+  parent.appendChild(errorMessage);
+}
+
+const fetchErrorHandling = async (response) => {
+  if (response.ok) {
+    return await response.json();
+  } else {
+    throw new Error("サーバーエラーが発生しました");
+  }
+}
+
+const getJsonOrError = async (url) => {
+  const response = await fetch(url);
+  const json = await fetchErrorHandling(response);
+  return json;
+}
+
+const getSlideImageData = async () => {
+  return new Promise((resolve) => {
+    setTimeout(() => resolve(getJsonOrError("https://myjson.dit.upm.es/api/bins/gzcn")), 3000);
+  })
+}
+
+const tryGetSlideImageData = async () => {
+  addLoading(slideWrapper);
+  try {
+    return await getSlideImageData();
+  } catch (e) {
+    addErrorMessage(e, slideWrapper);
+  } finally {
+    removeLoading();
+  }
+}
+
+const init = async () => {
+  const result = await tryGetSlideImageData();
+  if (result) {
+    createSlideItem(result);
+    createArrowButtons();
+  }
+}
+
+const createSlideItem = (imageData) => {
+  for (let i = 0; i < imageData.length; i++) {
+    const slideItem = createElementWithClassName("li", "slide-item");
+    const slideImage = createElementWithClassName("img", "slide-img");
+    slideImage.src = imageData[i].image;
+    // 初期設定として 最初の要素にis-displayingクラスを付与
+    if (i === 0) slideItem.classList.add("is-displaying");
+    slideItem.appendChild(slideImage);
+    slideList.appendChild(slideItem);
+  }
+}
```

Comment:
I had made same mistakes as you but Mr.Morita taught me that inserting and adding events are not suitable for `createSlideItem` because this function name is just creating. Now it works not only creating but also others in this function.
So I suggest you rename the function name or separate the insertion and event assignment from it. If you separate, this function just return the created element.

Reply:
Thanks for the review.
It's something I've received many times in my reviews

Modified the way createElement, appendChild, and event definitions... are divided into functions and return values are retrieved.

I defined init as a function to do the initialization, but I wonder if I'm including too many functions...
If you don't mind, please let me know what you think.

c6520a3

Reply:
A reviewer(@kanakonagiri) [advised ](https://github.com/sae-github/JS_lesson/pull/19#discussion_r759334535
e7e6bd5)me to make it a function. 
I'm modifying a few things in the init.
The functions for initializing slideitem and arrowbutton have been separated.

Please check this out as well.
e7e6bd5

Reply:
> I defined init as a function to do the initialization, but I wonder if I'm including too many functions...

I also thought same when I changed, so I understand your feeling.
But I think it become better than before. After functionalization, I think it's easier to understand.

I'm not sure this name is ok like `init**`, but for me, init is like a minimum setup that we have to do in order to do the requirements so I think it ok. I would like to ask others opinion here...🤔 
This is supplementary information but when I implemented it, I used named `render**` for the function witch creates and inserts element. And separated event assignment from this function. This idea is from [JavaScript best practices](https://www.w3.org/wiki/JavaScript_best_practices). Mr.Morita told me this website and this idea, so I would share with you too.

Reply:
Thanks for the reference article.
Naming was a point I was struggling with.
I was thinking of init** or setup** as a candidate.

I'd certainly like to hear other people's opinions here.